src/include/migraphx/op/loop.hpp

```diff
@@ -0,0 +1,164 @@
+        }
+
+        std::string out_prefix = "#output_";
+        loc                    = name.find(out_prefix);
+        if(loc != std::string::npos)
+        {
+            int index = std::stoi(name.substr(loc + out_prefix.size()));
+            return {index, false};
+        }
+
+        return {-1, false};
+    }
+
+    argument compute(const shape& out_shape,
+                     const std::vector<argument>& args,
+                     const std::vector<module_ref>& mods,
+                     const std::function<std::vector<argument>(
+                         module_ref&, const std::unordered_map<std::string, argument>&)>& run) const
+    {
+        auto iter_num = args.at(0).at<int64_t>();
+        if(iter_num > max_iter_num)
+        {
+            MIGRAPHX_THROW("LOOP compute(): iter_num " + std::to_string(iter_num) +
+                           " larger than max_iter_num " + std::to_string(max_iter_num) +
+                           ", please set the max_iter_num to at least " + std::to_string(iter_num) +
+                           " in onnx_options before parsing onnx file!");
+        }
+
+        auto cond                = args.at(1).at<bool>();
+        std::size_t dep_num      = args.size() - 2;
+        module_ref mod           = mods.at(0);
+        auto mod_name            = mod->name();
+        std::string param_prefix = "#" + mod->name() + "_in_";
+
+        std::vector<shape> vec_out_shapes = out_shape.sub_shapes();
+        std::vector<argument> scan_outputs;
+        std::transform(vec_out_shapes.begin() + dep_num,
+                       vec_out_shapes.end(),
+                       std::back_inserter(scan_outputs),
+                       [&](auto& s) { return argument{s}; });
+
+        std::vector<argument> in_args(args.begin() + 2, args.end());
+        shape s_iter{shape::int64_type};
+        shape s_cond{shape::bool_type};
+        uint64_t iter = 0;
+        in_args.insert(in_args.begin(), {s_cond, &cond});
+        in_args.insert(in_args.begin(), {s_iter, &iter});
+        const auto& param_name_shapes = mod->get_parameter_shapes();
+        for(iter = 0; (iter < iter_num) and cond; ++iter)
+        {
+            std::unordered_map<std::string, argument> params;
+            for(auto pn : param_name_shapes)
+            {
+                auto name     = pn.first;
+                auto io_index = get_name_index(name, param_prefix);
+                assert((io_index.first != -1) and (io_index.second));
+
+                // name is for input
+                if(io_index.second)
+                {
+                    params[name] = in_args.at(io_index.first);
+                }
+            }
+
+            auto mod_args = run(mod, params);
+            // copy loop carried dependency from mod outputs to inputs
+            std::copy(mod_args.begin(), mod_args.begin() + dep_num + 1, in_args.begin() + 1);
+            cond = mod_args.at(0).at<bool>();
+
+            // concat scan outputs
```

Comment:
If we are concating the output, I wonder if we can reuse the concat operator by maybe moving the implementation to a common function.